cloudsmith/data_source_package.go

```diff
@@ -50,6 +56,17 @@ func dataSourcePackageRead(d *schema.ResourceData, m interface{}) error {
 		}
 		d.Set("output_path", outputPath)
 		d.Set("output_directory", downloadDir)
+
+		// Calculate checksums for the downloaded file
+		md5Checksum, sha1Checksum, sha256Checksum, sha512Checksum, err := calculateChecksums(outputPath)
+		if err != nil {
+			return err
+		}
+
+		d.Set("checksum_md5", md5Checksum)
```

Comment:
To recap, the issues are:

1. If we download a file, it might be cached, and different from what the API says.
2. We're assuming the API is gospel and using/setting its response on the data source without checking.

Regardless of why, getting a checksum that differs from the API is the primary problem. What we should do here instead is to cache-bust this, but only if the download response was cached and the checksum is different.

Something like (pseudo-code):

```
outputPath, isCached, err := downloadPackage(pkg.GetCdnUrl(), downloadDir, pc)  // Added isCached return value
md5Checksum, sha1Checksum, sha256Checksum, sha512Checksum, err := calculateChecksums(outputPath)

checksumMismatch = false;

if (md5Checksum != pkg.GetChecksumMd5()) {  // and for sha1/sha256/sha512
  checksumMismatch = true;

  if (isCached) {
    // We're cached and the checksum differs, so bust the cache and try again 
    outputPath, isCached, err := downloadPackage(pkg.GetCdnUrl(), downloadDir, pc, true)  // Added a "bustCache" arg
    md5Checksum, sha1Checksum, sha256Checksum, sha512Checksum, err := calculateChecksums(outputPath)

    if (md5Checksum == pkg.GetChecksumMd5()) {  // and for sha1/sha256/sha512
      // We've now "fixed it"
      checksumMismatch = false;
    }    
  }

  if (checksumMismatch) {
    // We need to raise a hard error here that the checksum isn't correct, and explain which one (and maybe why)
  }
}
```

That would achieve the goal of fixing the underlying issue of the file being cached, still allow it to retrieve cached files assuming checksums are OK, retry if they're different, and ultimately error out if they're still broken after retry.

Reply:
Hey @lskillen - thanks for the feedback! 

2 questions:

1. What argument do I need to pass in order to bust the cache, would it be per your example from intercom by adding `?foo=1` ?
2. Instead of having a re-try logic, should we just pass that argument in the first call to bust the cache right away or would that hurt the performance?

Reply:
Great questions:

1. Let's use a unix/epoch timestamp, like `?time=1234567890`
2. It definitely would hurt performance to bust the cache every time; retry is on purpose. :)